File: cpp/tictactoe.cpp
#include <iostream>
#include <string>

struct player
{
    std::string token;
    std::string name;
    player(std::string _token, std::string _name) : token(_token), name(_name) {}
};

player player_1("O", "Player 1");
player player_2("X", "Player 2");

struct slot
{
    std::string empty_token = " ";
    std::string token = " ";

    bool update_token(std::string _token)
    {
        if(is_empty())
        {
            token = _token;
            return true;
        }
        return false;
    }

    bool is_empty()
    {
        return token == empty_token;
    }

    reset()
    {
        token = empty_token;
    }
};

struct board
{
    slot field[3][3];
    board()
    {
        for(int i = 0; i < 3; i++)
        {
            for(int j = 0; j < 3; j++)
            {
                slot s;
                field[i][j] = s;
            }
        }
    }

    int get_board_dimension_height()
    {
        return sizeof(field) / sizeof(field[0]);
    }

    int get_board_dimension_width()
    {
        return sizeof(field[0]) / sizeof(field[0][0]);
    }

    std::string draw(bool use_tokens)
    {
        std::string token;
        std::string printed = "";
        int height = get_board_dimension_height();
        int width = get_board_dimension_width();
        int ctr = 1;
        for(int i = 0; i < height; i++)
        {
            for(int j = 0; j < width; j++)
            {
                if(use_tokens)
                {
                    token = field[i][j].token;
                }
                else
                {
                    token = std::to_string(ctr);
                    ctr++;
                }

                if (j == 0)
                {
                    printed += "  " + token + " | ";
                }
                else
                {
                    printed += token;
                }

                if (0 < j && j < width - 1)
                {
                    printed += " | ";
                }
            }
            printed += "\n";
            if(i < height - 1)
            {
                printed += "-------------";
                printed += "\n";
            }
        }
        return printed;
    }

    bool write_turn(player _player, int input)
    {
        int ctr = 1;
        for(int i = 0; i < get_board_dimension_width(); i++)
        {
            for(int j = 0; i < get_board_dimension_height(); j++)
            {
                if(input == ctr)
                {
                    return field[i][j].update_token(_player.token);
                }
                else
                {
                    ctr++;
                }
            }
        }
        return false;
    }

    bool equal_all_slots(slot arr[3])
    {
        bool equal = true;
        for(int i = 1; i < 3; i++)
        {
            slot current = arr[i];
            slot previous = arr[i - 1];
            if(current.token == previous.token && !current.is_empty() && !previous.is_empty())
            {
                continue;
            }
            else
            {
                equal = false;
            }
        }
        return equal;
    }

    bool check_win()
    {
        bool win = false;

        int width = get_board_dimension_width();
        int height = get_board_dimension_height();
        for(int i = 0; i < get_board_dimension_width(); i++)
        {
            slot horizontal_slots[width];
            slot vertical_slots[width];
            // Check horizontal and vertical.
            for(int j = 0; j < get_board_dimension_height(); j++)
            {
                slot h_slot;
                slot v_slot;
                h_slot = field[i][j];
                v_slot = field[j][i];
                horizontal_slots[j] = h_slot;
                vertical_slots[j] = v_slot;
            }

            bool h_win = equal_all_slots(horizontal_slots);
            bool v_win = equal_all_slots(vertical_slots);
            if(h_win || v_win)
            {
                win = true;
                break;
            }
        }

        // Check diagonal.
        slot d1[] = {field[0][0], field[1][1], field[2][2]};
        slot d2[] = {field[0][2], field[1][1], field[2][0]};
        bool d1_win = equal_all_slots(d1);
        bool d2_win = equal_all_slots(d2);
        if(d1_win || d2_win)
        {
            win = true;
        }

        return win;
    }

    bool is_full()
    {
        int width = get_board_dimension_width();
        int height = get_board_dimension_height();
        int filled;
        for(int i = 0; i < width; i++)
        {
            for(int j = 0; j < height; j++)
            {
                if(!field[i][j].is_empty())
                {
                    filled++;
                }
            }
        }
        return filled == width * height;
    }

    reset()
    {
        int width = get_board_dimension_width();
        int height = get_board_dimension_height();
        for(int i = 0; i < width; i++)
        {
            for(int j = 0; j < height; j++)
            {
                field[i][j].reset();
            }
        }
    }
};

player toggle_player(player p)
{
    if(p.token == player_1.token)
    {
        return player_2;
    }
    else
    {
        return player_1;
    }
}

clear_screen()
{
    // Make sure system is safe to use before clear screen.
    if(system( NULL )) system("CLS");
}

main()
{
    board game_board;
    player current_player = player_1;
    bool win = false;
    bool draw = false;
    bool again = true;
    std::string alert_message = "";

    while(!win && !draw && again)
    {
        clear_screen();

        std::cout << "Guide\n";
        std::cout << game_board.draw(false);
        std::cout << "\n\n";

        std::cout << "Playing Field\n";
        std::cout << game_board.draw(true) << std::endl;
        std::cout << alert_message;
        std::cout << "Turn of " << current_player.name << " [" << current_player.token << "]\n";

        int input;
        std::cout << "Input: ";
        std::cin >> input;

        bool valid_move = game_board.write_turn(current_player, input);
        win = game_board.check_win();
        if(valid_move)
        {
            if(!win)
            {
                current_player = toggle_player(current_player);
            }
            alert_message = "";
        }
        else
        {
            alert_message = "Slot " + std::to_string(input) + " is either taken or invalid.\n\n";
        }

        if(!win && game_board.is_full())
        {
            draw = true;
        }

        if(win || draw)
        {
            clear_screen();
            std::cout << game_board.draw(true);
            if(win)
            {
                std::cout << current_player.name << " has won!\n\n";
            }
            if(draw)
            {
                std::cout << "Draw!\n\n";
            }

            std::string play_again;
            std::cout << "Play again? (Y/N) : ";
            std::cin >> play_again;

            if(play_again == "Y" || play_again == "y")
            {
                win = false;
                again = true;
                draw = false;
                game_board.reset();
            }
            else
            {
                win = false;
                again = false;
                draw = false;
            }
        }
    }
}

File: csharp/tictactoe.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using tictactoe;

namespace tictactoe
{
    /// <summary>
    /// A class representing a player.
    /// </summary>
    public class Player
    {
        private string token;
        private string name;

        /// <summary>
        /// Create a player.
        /// </summary>
        public Player(string token, string name)
        {
            this.token = token;
            this.name = name;
        }

        /// <summary>
        /// The readonly property for the player's name.
        /// </summary>
        public string Name
        {
            get { return name; }
        }

        /// <summary>
        /// The readonly property for the player's token.
        /// </summary>
        public string Token
        {
            get { return token; }
        }
    }

    /// <summary>
    /// A class representing a grid in the playing field.
    /// </summary>
    public class Slot
    {
        private string token;
        private Regex emptyRegex;

        /// <summary>
        /// Create a slot.
        /// </summary>
        public Slot(string token)
        {
            this.token = token;
            emptyRegex = new Regex(@"^\s*$");
        }

        /// <summary>
        /// The token in the grid.
        /// </summary>
        public string Token
        {
            get { return token; }
            set { this.token = value; }
        }

        /// <summary>
        /// Checks if the slot contains a token.
        /// </summary>
        /// <returns>
        /// Returns true if the slot is empty.
        /// If the token only contains spaces,
        /// it counts as empty.
        /// </returns>
        public bool IsEmpty()
        {
            return emptyRegex.Matches(token).Count > 0;
        }

        /// <summary>
        /// Reset the token as a string with a single space.
        /// </summary>
        public void Reset()
        {
            token = " ";
        }
    }


    /// <summary>
    /// A class representing the playing field for Tic-Tac-Toe.
    /// </summary>
    public class Board
    {
        private Slot[,] field;
        private const string initialToken = " ";
        private int height;
        private int width;


        /// <summary>
        /// Create a board.
        /// </summary>
        public Board(int height, int width)
        {
            field = new Slot[height, width];
            for(int i = 0; i < height; i++)
            {
                for(int j = 0; j < width; j++)
                {
                    field[i, j] = new Slot(initialToken);
                }
            }
            this.height = height;
            this.width = width;
        }

        /// <summary>
        /// Checks if the board is already full of tokens. Returns a boolean value.
        /// </summary>
        /// <returns>
        /// Returns true if all slots in the board are not empty.
        /// </returns>
        public bool IsFull()
        {
            int filled = 0;
            for(int i = 0; i < height; i++)
                for(int j = 0; j < width; j++)
                    filled += field[i, j].IsEmpty() ? 0 : 1;
            return filled == width * height;
        }

        /// <summary>
        /// Draws the board in the console.
        /// </summary>
        /// <param name="useTokens">
        /// If false, print the board in guide version by representing each slot
        /// with numbers corresponding to the input required to insert a token there.
        /// If true, print the board as is with the tokens currently in it.
        /// </param>
        /// <returns>
        /// Returns a string representing the current status of board. If usetokens
        /// us false, then the board will be printed in guide mode.
        /// </returns>
        public string Draw(bool useTokens)
        {
            List<string> printed = new List<string>();
            int ctr = 1;
            for(int i = 0; i < height; i++)
            {
                for(int j = 0; j < width; j++)
                {
                    string token = useTokens ? field[i, j].Token : ctr.ToString();
                    ctr += useTokens ? 0 : 1;
                    printed.Add(j == 0 ? " " + token + " | " : token);
                    if(0 < j && j < width - 1)
                        printed.Add(" | ");
                }
                printed.Add(i < height - 1 ? "\n-------------\n" : "\n");
            }
            return string.Join("", printed);
        }

        /// <summary>
        /// Writes player's move into the board.
        /// </summary>
        /// <param name="player">
        /// The player making the move.
        /// </param>
        /// <param name="move">
        /// The value of the player's input as the move.
        /// </param>
        /// <returns>
        /// Returns true if the turn is written successfully.
        /// </returns>
        public bool WriteTurn(Player player, string move)
        {
            int ctr = 1;
            for(int i = 0; i < width; i++)
            {
                for(int j = 0; j < height; j++)
                {
                    int moveNumber;
                    try
                    {
                        moveNumber = Int32.Parse(move);
                    }
                    catch(FormatException)
                    {
                        return false;
                    }
                    if(moveNumber == ctr)
                    {
                        bool empty = field[i, j].IsEmpty();
                        if(empty)
                            field[i, j].Token = player.Token;
                        return empty;
                    }
                    else
                    {
                        ctr++;
                    }
                }
            }
            return false;
        }


        /// <summary>
        /// Checks if all values from a collection of slots are equal.
        /// </summary>
        /// <param name="collection">
        /// A List of slots.
        /// </param>
        /// <returns>
        /// Returns true if all are equal and not empty.
        /// </returns>
        private bool CheckAllEqual(List<Slot> collection)
        {
            foreach(Slot current in collection)
                if(!(current.Token == collection[0].Token && !current.IsEmpty()))
                    return false;
            return true;
        }


        /// <summary>
        /// Checks if a win condition is already met.
        /// </summary>
        /// <returns>
        /// Returns true if a win condition is met.
        /// </returns>
        public bool CheckWin()
        {
            bool win = false;
            for(int i = 0; i < width; i++)
            {
                List<Slot> horizontalSlots = new List<Slot>();
                List<Slot> verticalSlots = new List<Slot>();
                List<Slot> diagonalTopToBottomSlots = new List<Slot>();
                List<Slot> diagonalBottomToTopSlots = new List<Slot>();

                /* Check horizontal and vertical. */
                for(int j = 0; j < height; j++)
                {
                    horizontalSlots.Add(field[i, j]);
                    verticalSlots.Add(field[j, i]);
                }

                int h = height - 1;
                /* Check diagonals. */
                for(int j = 0; j < height; j++)
                {
                    diagonalTopToBottomSlots.Add(field[j, j]);
                    diagonalBottomToTopSlots.Add(field[h, j]);
                    diagonalBottomToTopSlots.Add(field[j, h]);
                    h--;
                }

                win = (
                    CheckAllEqual(verticalSlots) ||
                    CheckAllEqual(horizontalSlots) ||
                    CheckAllEqual(diagonalTopToBottomSlots) ||
                    CheckAllEqual(diagonalBottomToTopSlots)
                );
                if(win)
                    break;
            }
            return win;
        }

        /// <summary>
        /// Resets the entire board for another game.
        /// </summary>
        public void Reset()
        {
            for(int i = 0; i < this.width; i++)
                for(int j = 0; j < this.height; j++)
                    field[i, j].Reset();
        }
    }
}
